Add Imperative Slot API

yuzhe-han opened the pull request:
<!--
Thank you for contributing to the HTML Standard! Please describe the change you are making and complete the checklist below if your change is not editorial.
-->
The explainer for this feature is here: https://github.com/w3c/webcomponents/blob/gh-pages/proposals/Imperative-Shadow-DOM-Distribution-API.md  
The issue discussion is here: [3534](https://github.com/whatwg/html/issues/3534)  
There is a corresponding [Pull Request](https://github.com/whatwg/dom/pull/860) for the DOM spec that goes along with this PR.  

- [ ] At least two implementers are interested (and none opposed):
   * I'm working to get this implemented in Chromium, tracking bug [here](http://crbug.com/869308).
   * The implementation is based on what was agreed in the last [F2F Meeting](https://github.com/whatwg/html/issues/3534#issuecomment-537802687).

- [ ] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * In WPT: https://wpt.fyi/results/shadow-dom/slots-imperative-slot-api.tentative.html?label=master&label=experimental&aligned&q=slots-imperative-slot-api.tentative.html
   * In WPT: https://wpt.fyi/results/shadow-dom/slots-imperative-api-slotchange.tentative.html?label=master&label=experimental&aligned
   * More tests are being implemented, but the basics are there.
- [x] [Implementation bugs](https://github.com/whatwg/meta/blob/master/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chrome: https://crbug.com/869308
   * Firefox: None yet
   * Safari: None yet


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/5483/acknowledgements.html" title="Last updated on Apr 7, 2021, 6:14 PM UTC (fdc36ca)">/acknowledgements.html</a>  ( <a href="https://whatpr.org/html/5483/e9d38e8...fdc36ca/acknowledgements.html" title="Last updated on Apr 7, 2021, 6:14 PM UTC (fdc36ca)">diff</a> )
<a href="https://whatpr.org/html/5483/infrastructure.html" title="Last updated on Apr 7, 2021, 6:14 PM UTC (fdc36ca)">/infrastructure.html</a>  ( <a href="https://whatpr.org/html/5483/e9d38e8...fdc36ca/infrastructure.html" title="Last updated on Apr 7, 2021, 6:14 PM UTC (fdc36ca)">diff</a> )
<a href="https://whatpr.org/html/5483/scripting.html" title="Last updated on Apr 7, 2021, 6:14 PM UTC (fdc36ca)">/scripting.html</a>  ( <a href="https://whatpr.org/html/5483/e9d38e8...fdc36ca/scripting.html" title="Last updated on Apr 7, 2021, 6:14 PM UTC (fdc36ca)">diff</a> )

annevk commented on source:
```diff
@@ -58583,6 +58586,7 @@ interface <dfn>HTMLSlotElement</dfn> : <span>HTMLElement</span> {
   [<span>CEReactions</span>] attribute DOMString <span data-x="dom-slot-name">name</span>;
   sequence&lt;Node> <span data-x="dom-slot-assignedNodes">assignedNodes</span>(optional <span>AssignedNodesOptions</span> options = {});
   sequence&lt;Element> <span data-x="dom-slot-assignedElements">assignedElements</span>(optional <span>AssignedNodesOptions</span> options = {});
+  void <span data-x="dom-slot-assign">assign</span>(sequence&lt;Node> nodes);
```
Should this be `Node... nodes` or even `(Element or Text)... nodes`? This would match more recent DOM APIs.

yuzhe-han replied:
Thanks annevk@. I've updated PR based on your suggestion.

annevk commented on source:
```diff
@@ -58655,6 +58665,21 @@ dictionary <dfn>AssignedNodesOptions</dfn> {
+  <ol>
+   <li><p>Check <span>slot</span>'s <span>shadow root</span> is set to "manual"
+   <span>slot assignment</span>. Otherwise, throw <span>"<code>NotAllowedError</code>"</span>
+   <code>DOMException</code>.</p></li>
```
If slot's shadow root's slot assignment' is not "manual", then throw ... would be more natural. Assuming that's what it's trying to say.

yuzhe-han replied:
Thanks for the suggestion. It's clearer and I've updated my PR.

annevk commented on source:
```diff
@@ -58655,6 +58665,21 @@ dictionary <dfn>AssignedNodesOptions</dfn> {
+   <span>shadow host</span>. Otherwise, throw <span>"<code>NotAllowedError</code>"</span>
+   <code>DOMException</code>.</p></li>
+
+   <li><p>Run <span>assign slottables</span> for <span data-x="concept-slot">slot</span></p></li>
```
s/slot/this/, right?

yuzhe-han replied:
Right!

yuzhe-han replied:
Appreciated your quick review!

annevk commented:
As far as I know you want to track which nodes are manually assigned. You need some infrastructure for that which isn't provided/hooked into here.

yuzhe-han commented:
@annevk,

I updated the spec to include manuallyAssignedNodes concept. 
I also fetched from whatwg:master to update my branch and rebased my changes, not sure if that's advised since  the rebase modified the reviews you did earlier.

Please take a look,  Thanks.

sideshowbarker commented on source:
```diff
@@ -58553,6 +58564,31 @@ dictionary <dfn>AssignedNodesOptions</dfn> {
+
+   <li><p>Run <span>assign slottables for a tree</span> for <b>this</b>.</p></li>
+
+   <p class="note">It would be worth noting that <dfn data-export="">manuallyAssignedNodes</dfn>
```
A normative term can’t be defined in a Note, because Notes are non-normative. Also it seems *“It would be worth noting that manuallyAssignedNodes is not used as assigned nodes as is.”* isn’t actually defining `manuallyAssignedNodes` but instead just making a statement about it. And to me at least, I can’t parse the phrase *”manuallyAssignedNodes is not used as assigned nodes as is”* — it’s unclear what it’s meant to be stating. Regardless, it seems like a normative definition for `manuallyAssignedNodes` still needs to be added to this patch.

annevk replied:
Indeed, this does not work.

yuzhe-han replied:
Understood.  I've removed this section and defined manuallyAssignedNodes on top of assign(nodes) function.

annevk commented on source:
```diff
@@ -58553,6 +58564,31 @@ dictionary <dfn>AssignedNodesOptions</dfn> {
+  when invoked, must run these steps:</p>
+
+  <ol>
+   <li><p>If <span>slot</span>'s <span>shadow root</span> <span>slot assignment</span> is not
```
shadow root's

annevk commented on source:
```diff
@@ -58553,6 +58564,31 @@ dictionary <dfn>AssignedNodesOptions</dfn> {
+  when invoked, must run these steps:</p>
+
+  <ol>
+   <li><p>If <span>slot</span>'s <span>shadow root</span> <span>slot assignment</span> is not
```
Which slot? Did you mean `<span>this</span>`? What if it doesn't have a shadow root?

yuzhe-han replied:
Updated spec to include a case for none shadow root and specified it's 'this' slot.

annevk commented on source:
```diff
@@ -58553,6 +58564,31 @@ dictionary <dfn>AssignedNodesOptions</dfn> {
+   <li><p>If <span>slot</span>'s <span>shadow root</span> <span>slot assignment</span> is not
+   "manual", then throw <span>"<code>NotAllowedError</code>"</span> <code>DOMException</code>.</p></li>
+
+   <li><p>Check each node in nodes is a <span>slottable</span> child of <span>slot</span>'s
```
For each _node_ of _nodes_: if _node_ is not a child (needs xref) of slot (which slot?)'s shadow host. What does the slottable check add?

yuzhe-han replied:
Updated to include child (xref) and this slot. Your right, slottable isn't needed since the last step, assign slottables for a tree, takes care of that.

annevk commented on source:
```diff
@@ -58553,6 +58564,31 @@ dictionary <dfn>AssignedNodesOptions</dfn> {
+
+   <li><p>For each <var>node</var> in <var>nodes</var>, remove <var>node</var> from the
+   <span>manuallyAssignedNodes</span> of each <span>slot</span> in this <span>shadow root</span>’s
+   <span>inclusive descendants</span>, if it exists.
```
What if the node was assigned to other slots in other shadow trees?

yuzhe-han replied:
The previous step will throw an NotAllowedError, if the node is not a child of this shadow host.

yuzhe-han commented:
Thanks for reviewing. I've updated the spec based on feedback.

yuzhe-han commented on source:
```diff
@@ -58553,6 +58564,30 @@ dictionary <dfn>AssignedNodesOptions</dfn> {
+   <code>DOMException</code>.</p></li>
+
+   <li><p>For each <var>node</var> in <var>nodes</var>: if <var>node</var> is not a <span data-x="concept-tree-child">child</span>
+   of <span>this</span> <span>shadow host</span>, then throw <span>"<code>NotAllowedError</code>"</span>
```
Right.

yuzhe-han commented on source:
```diff
@@ -58553,6 +58564,30 @@ dictionary <dfn>AssignedNodesOptions</dfn> {
+   <span>shadow root</span>'s <span>slot assignment</span> is not "manual", then throw <span>"<code>NotAllowedError</code>"</span>
+   <code>DOMException</code>.</p></li>
+
+   <li><p>For each <var>node</var> in <var>nodes</var>: if <var>node</var> is not a <span data-x="concept-tree-child">child</span>
```
You are right. The algorithms in corresponding DOM spec PR doesn't update manually assigned nodes when an assigned node or a slot is removed.

I looked at the #concept-node-remove to see how when "slot assignment' auto is handled. 
Step 11 and 12 takes care of the assigned node case[1]. And step 14 takes care of the case[2] if the removed node is a slot or it contains a slot descendant.

So in theory, when "slot assignment" manual, it can use the same algorithm to make sure that slottables at the 3 of #assign-slotables are set back to manually assigned nodes. Thus, I updated the corresponding DOM spec to keep slot's assigned nodes and manually assigned nodes in sync. 

Hopefully, this works. I also looked at the possibility of clearing out the manually assigned nodes during #concept-node-remove, but decided to go with a similar path for auto assignment.

annevk commented:
I wonder if it wouldn't be easier if instead the children of a host had a reference to null or a slot. Then whenever you need the "assigned slottables" of a slot you can go through those children and if you need to update anything you can go through those children rather than traversing the entire shadow tree looking for slots. And if you compute the "assigned slottables" lazily I guess you only need to care for mutations to the host's children, potentially simplifying the changes in DOM.

cc @smaug---- @rniwa

smaug---- commented:
> had a reference to null or a slot.

Current DOM's https://dom.spec.whatwg.org/#find-a-slot is going through the tree.
(Though Gecko does keep element->slot reference to have fast access)

yuzhe-han commented:
Anne@, 
Thanks for your suggestion. I think for children of a host to have a reference to either null or slot makes sense. I send you an email for some clarification on how to best move forward. 

Thanks.

rniwa commented:
FWIW, what's being proposed here seems fine.

annevk commented on source:
```diff
@@ -58553,6 +58564,29 @@ dictionary <dfn>AssignedNodesOptions</dfn> {
   </ol>
 
+  <p>The <code>slot</code> element has a <dfn data-dfn-for="HTMLSlotElement" data-export="">manually assigned nodes</dfn>,
+  which is a list of <span>slottable</span>s set by <code data-x="dom-slot-assign">assign(<var>nodes</var>)</code>.</p>
```
* s/list/set/
* Use `<span data-x=slottable>slottables</span>`
* There's no need to say here what it is set by, but maybe that's okay. I'll leave that to Domenic.

annevk commented on source:
```diff
@@ -58553,6 +58564,29 @@ dictionary <dfn>AssignedNodesOptions</dfn> {
+  when invoked, must run these steps:</p>
+
+  <ol>
+   <li><p>If <span>this</span> <span>root</span> is not a <span>shadow root</span> or its <span>slot assignment</span>
```
`<span>this</span>'s`

annevk commented on source:
```diff
@@ -58553,6 +58564,29 @@ dictionary <dfn>AssignedNodesOptions</dfn> {
+
+  <ol>
+   <li><p>If <span>this</span> <span>root</span> is not a <span>shadow root</span> or its <span>slot assignment</span>
+   is not "manual", then throw <span>"<code>NotAllowedError</code>"</span> <code>DOMException</code>.</p></li>
```
manual should be in `<code data-x="">`

annevk commented on source:
```diff
@@ -58553,6 +58564,29 @@ dictionary <dfn>AssignedNodesOptions</dfn> {
+   is not "manual", then throw <span>"<code>NotAllowedError</code>"</span> <code>DOMException</code>.</p></li>
+
+   <li><p>For each <var>node</var> in <var>nodes</var>: if <var>node</var> is not a <span data-x="concept-tree-child">child</span>
+   of <span>this</span> <span>root</span>'s <span>shadow host</span>, then throw <span>"<code>NotAllowedError</code>"</span>
```
this's again

annevk commented on source:
```diff
@@ -58553,6 +58564,29 @@ dictionary <dfn>AssignedNodesOptions</dfn> {
+   of <span>this</span> <span>root</span>'s <span>shadow host</span>, then throw <span>"<code>NotAllowedError</code>"</span>
+   <code>DOMException</code>.</p></li>
+
+   <li><p>For each <var>node</var> in <var>nodes</var>, remove <var>node</var> from the
```
I think you want `<span data-x="set remove">remove</span>` here.

yuzhe-han replied:
I didn't see 'set remove', but used 'list remove' instead.

annevk commented on source:
```diff
@@ -58553,6 +58564,29 @@ dictionary <dfn>AssignedNodesOptions</dfn> {
+   <code>DOMException</code>.</p></li>
+
+   <li><p>For each <var>node</var> in <var>nodes</var>, remove <var>node</var> from the
+   <span>manually assigned nodes</span> of each <span>slot</span> in this <span>shadow root</span>’s
```
Do you mean this's root's descendants here?

yuzhe-han replied:
Yes.  Updated. Thanks.

annevk commented on source:
```diff
@@ -58553,6 +58564,29 @@ dictionary <dfn>AssignedNodesOptions</dfn> {
+
+   <li><p>Set <span>this</span> <span>manually assigned nodes</span> to <var>nodes</var>.</p></li>
+
+   <li><p>Run <span>assign slottables for a tree</span> for <b>this</b>.</p></li>
```
`<span>this</span>`

annevk reviewed (commented):
I think this is good now, except for the nits. It does seem as defined `assign()` is quite expensive, but also has lots of room for optimizations.

yuzhe-han commented:
@annevk,

Thanks for reviewing.  
If it makes it easier, I can be available for reviewing this together. 

Thanks.

annevk commented on source:
```diff
@@ -59388,6 +59399,29 @@ dictionary <dfn>AssignedNodesOptions</dfn> {
    to contain only <code>Element</code> nodes.</p></li>
   </ol>
 
+  <p>The <code>slot</code> element has a <dfn data-dfn-for="HTMLSlotElement" data-export="">manually assigned nodes</dfn>,
```
s/a//

yuzhe-han replied:
done.

annevk commented on source:
```diff
@@ -59388,6 +59399,29 @@ dictionary <dfn>AssignedNodesOptions</dfn> {
+
+  <ol>
+   <li><p>If <span>this</span>'s <span>root</span> is not a <span>shadow root</span> or its <span>slot assignment</span>
+   is not <code data-x="">manual</code>, then throw <span>"<code>NotAllowedError</code>"</span> <code>DOMException</code>.</p></li>
```
throw a*

annevk commented on source:
```diff
@@ -59388,6 +59399,29 @@ dictionary <dfn>AssignedNodesOptions</dfn> {
+   is not <code data-x="">manual</code>, then throw <span>"<code>NotAllowedError</code>"</span> <code>DOMException</code>.</p></li>
+
+   <li><p>For each <var>node</var> in <var>nodes</var>: if <var>node</var> is not a <span data-x="concept-tree-child">child</span>
+   of <span>this</span>'s <span>root</span>'s <span>shadow host</span>, then throw <span>"<code>NotAllowedError</code>"</span>
```
throw a*

yuzhe-han replied:
done

annevk commented on source:
```diff
@@ -59388,6 +59399,29 @@ dictionary <dfn>AssignedNodesOptions</dfn> {
+
+   <li><p>For each <var>node</var> in <var>nodes</var>, <span data-x="list remove">remove</span> <var>node</var>
+   from the <span>manually assigned nodes</span> of each <span>slot</span> in <span>this</span>'s <span>root</span>’s
+   descendants, if it exists.</p></li>
```
xref descendant.

yuzhe-han replied:
done.

annevk commented on source:
```diff
@@ -59388,6 +59399,29 @@ dictionary <dfn>AssignedNodesOptions</dfn> {
+
+   <li><p>Set <span>this</span> <span>manually assigned nodes</span> to <var>nodes</var>.</p></li>
+
+   <li><p>Run <span>assign slottables for a tree</span> for <span>this</span>.</p></li>
```
Shouldn't you pass this's root here?

yuzhe-han replied:
Right!  Done.

rniwa commented:
https://bugs.webkit.org/show_bug.cgi?id=218692

annevk commented on source:
```diff
@@ -59364,6 +59376,29 @@ dictionary <dfn>AssignedNodesOptions</dfn> {
+   from the <span>manually assigned nodes</span> of each <span>slot</span> in <span>this</span>'s <span>root</span>’s
+   <span data-x="concept-tree-descendant">descendants</span>, if it exists.</p></li>
+
+   <li><p>Set <span>this</span> <span>manually assigned nodes</span> to <var>nodes</var>.</p></li>
```
```suggestion
   <li><p>Set <span>this</span>'s <span>manually assigned nodes</span> to <var>nodes</var>.</p></li>
```

annevk commented on source:
```diff
@@ -59337,6 +59343,12 @@ dictionary <dfn>AssignedNodesOptions</dfn> {
+   <dd>Set the list of nodes as <code data-x="dom-slot-assignedNodes">assignedNodes</code>.
+   <p>Throws a <span>"<code>NotAllowedError</code>"</span> <code>DOMException</code>
+   if <span>slot</span>'s <span>shadow root</span> is not "manual" <span>slot assignment</span>
+   or nodes are not <span>slottable</span> children of its <span>shadow host</span>.</dd>
```
This should be cleaned up a bit. The `<p>` comes out of nowhere, `assignedNodes` is a method, shadow root is not a property of a slot...

yuzhe-han replied:
updated.

tetsuharuohzeki commented on source:
```diff
@@ -59292,6 +59297,7 @@ interface <dfn>HTMLSlotElement</dfn> : <span>HTMLElement</span> {
   [<span>CEReactions</span>] attribute DOMString <span data-x="dom-slot-name">name</span>;
   sequence&lt;Node> <span data-x="dom-slot-assignedNodes">assignedNodes</span>(optional <span>AssignedNodesOptions</span> options = {});
   sequence&lt;Element> <span data-x="dom-slot-assignedElements">assignedElements</span>(optional <span>AssignedNodesOptions</span> options = {});
+  void <span data-x="dom-slot-assign">assign</span>((<span>Element</span> or <code>Text</code>)... nodes);
```
This should be `undefined`. Others in the spec are updated to it by  https://github.com/whatwg/html/pull/5838

annevk commented:
@domenic @rniwa can either of you give this and the DOM PR at https://github.com/whatwg/dom/pull/860 a review? I have some nits that I can address, but I'm worried about losing the plot given the amount of time that has passed.

domenic reviewed (approved):
I pushed some tweaks. The only one of substance is that I made it explicit that you loop through descendants in tree order when removing them. I imagine the order might be observable using mutation observers?

I also tidied some nearby method definitions and IDL stuff to conform to modern conventions.

The HTML side of this seems relatively simple, so next up is the DOM side...

domenic commented:
Hmm, I also changed "manually assigned nodes" from a set to a list, because I didn't see any protection against `el.assign(node, node)`. Is that change OK, @yuzhe-han?

yuzhe-han commented:
@domenic, 
  Thanks for your clean up on this PR.
  Your change from a set to a list look correct to me.

domenic commented:
OK, cool. So there's no problem with the algorithms if the same node appears twice in the manually assigned nodes list? Do we have test cases for the behavior in that scenario?

rniwa commented on source:
```diff
@@ -59295,33 +59301,79 @@ dictionary <dfn>AssignedNodesOptions</dfn> {
+    </ol>
+   </li>
+
+   <li><p>Set <span>this</span>'s <span>manually assigned nodes</span> to <var>nodes</var>.</p></li>
```
This isn't quite right. We need to remove duplicates from the manually assigned nodes since "find slottables" in https://github.com/whatwg/dom/pull/860/files doesn't remove duplicates either.

domenic replied:
Thanks, that seems more correct to me, and seems to match the implementation in Chromium. https://source.chromium.org/chromium/chromium/src/+/master:third_party/blink/renderer/core/html/html_slot_element.cc;l=187;drc=510d6b8791a0e20c3470f9c0fcf1472d1bb71995

I will update the CL.

annevk replied:
It seems slightly cleaner to make the set during the first iteration to avoid redundant descendant searches, but I think it should not be observable...

domenic replied:
I did update the PR; see commit https://github.com/whatwg/html/pull/5483/commits/fdc36cad40089f53d89ff286efffff06b459e106

annevk commented on source:
```diff
@@ -59295,33 +59301,85 @@ dictionary <dfn>AssignedNodesOptions</dfn> {
+  <p>The <code>slot</code> element has <dfn export for="HTMLSlotElement">manually assigned
+  nodes</dfn>, which is an <span data-x="set">ordered set</span> of <span
+  data-x="slottable">slottables</span> set by <code data-x="dom-slot-assign">assign()</code>. It is
+  initially empty.</p>
```
Should we move this up? Usually we describe "slots" before public members.

rniwa commented on source:
```diff
@@ -59295,33 +59301,85 @@ dictionary <dfn>AssignedNodesOptions</dfn> {
+   <li><p><span data-x="list iterate">For each</span> <var>node</var> of <var>nodes</var>: if
+   <var>node</var> is not a <span data-x="concept-tree-child">child</span> of <span>this</span>'s
+   <span>root</span>'s <span>shadow host</span>, then throw a
+   <span>"<code>NotAllowedError</code>"</span> <code>DOMException</code>.</p></li>
```
Hm... what if you wanted to assign nodes then add them as children of the shadow host? The one of main use cases of manual assignment is so that custom element can have JS API which accepts a list of nodes to use, right? In that scenario, it's possible for the users of a custom element to create a bunch of nodes, call JS API, then add them as children of the host instead of adding them first and then giving them to the JS API. At least, that was one of the use cases that came up during accessibility related discussion in the past which motivated the current design.

mfreed7 reviewed (commented):
I've opened another PR for this, with the changes requested: https://github.com/whatwg/html/pull/6561

smaug---- commented:
So did we get any tests and does any implementation pass them?
It seems like the implementation in Blink
https://source.chromium.org/chromium/chromium/src/+/master:third_party/blink/renderer/core/html/html_slot_element.idl;l=35
is quite different to the spec https://html.spec.whatwg.org/#htmlslotelement
(see different params to assign())
